### ISSUE TITLE
Submitting patch to read GML geo metadata and other associated data for inclusion in openjpeg

### ISSUE THREAD

**Issue by HWiman** [uclouvain/openjpeg#1109]
We need the GML meta data stored in some JP2 files as described [here](http://docs.opengeospatial.org/is/08-085r4/08-085r4.html). 

There was no reader for ASOC, LBL or XML boxes so I have added that (v2.3.0). I hope you will consider adding these changes to future versions of openjpeg. I submit the diff patches for review. Please let me know if I can assist in getting this into openjpeg. 

The patches add a new `opj_jp2_asoc_t` type. An `opj_jp2` decoder has `numasoc` number of such instances in `*asoc`. For retrieval on client side, these data are copied to `opj_codestream_info_v2_t` for direct access and can be dumped using `opj_dump_associated_data( opj_codestream_info_v2_t* cstr_info,  FILE *output_stream )`.

For example the dump of a Sentinel-2 scene may look like attached dump.txt.
[openjpeg.c.diff.txt](https://github.com/uclouvain/openjpeg/files/1828822/openjpeg.c.diff.txt)
[openjpeg.h.diff.txt](https://github.com/uclouvain/openjpeg/files/1828823/openjpeg.h.diff.txt)
[jp2.c.diff.txt](https://github.com/uclouvain/openjpeg/files/1828824/jp2.c.diff.txt)
[jp2.h.diff.txt](https://github.com/uclouvain/openjpeg/files/1828826/jp2.h.diff.txt)
[dump.txt](https://github.com/uclouvain/openjpeg/files/1828829/dump.txt)

/Håkan Wiman

**Comment by DavidUnderdown** [comment]
In the meantime you can always use [exiftool](https://www.sno.phy.queensu.ca/~phil/exiftool/) to read/write XML from JP2s

**Comment by HWiman** [comment]
Yes thanks, but we want to keep the number of external dependencies low.

**Comment by boxerab** [comment]
Can you make a pull request, instead of attaching diff files ?
